src/elements/play-assets/file-upload-dropper.ts

```diff
@@ -0,0 +1,268 @@
+    if (changedProperties.has('acceptTypes')) {
+      this._accept = flattenAcceptTypes(this.acceptTypes)
+      this.#acceptPatterns = this.acceptTypes.flatMap(type => {
+        return Object.keys(type.accept).flatMap(mimetype => {
+          if (mimetype.includes('*')) {
+            return new RegExp(mimetype.split('*').join('.*'))
+          }
+          this.#allowedMimes.push(mimetype)
+          return []
+        })
+      })
+    }
+  }
+
+  protected override render(): TemplateResult {
+    return html`
+      <input
+        type="file"
+        id="fileInput"
+        accept="${ifDefined(this._accept)}"
+        ?multiple="${this.multiple}"
+        @change=${this.#processFileInput}
+      />
+      <label for="${PlayAssets.hasFileAccessAPI ? '' : 'fileInput'}">
+        <div
+          id="fileDrop"
+          class="${classMap({dragenter: this._dragging})}"
+          @click=${PlayAssets.hasFileAccessAPI ? this.#pickFile : undefined}
+          @dragenter=${this.#dragStart}
+          @dragover=${this.#dragStart}
+          @dragleave=${this.#dragEnd}
+          @dragend=${this.#dragEnd}
+          @drop=${this.#processDrop}
+        >
+          <slot></slot>
+          <span id="errorMessage">${this._errorMessage}</span>
+        </div>
+      </label>
+    `
+  }
+
+  /**
+   * Uses File Access API to get FileSystemFileHandles
+   */
+  #pickFile = async () => {
+    const fileHandles = await fileAccessContext.showOpenFilePicker({
+      ...(this.id ? {id: this.id} : {}),
+      types: this.acceptTypes,
+      multiple: this.multiple
+    })
+    if (fileHandles.length > 0) {
+      this.dispatchEvent(
+        Bubble<FilesSelectedEvent>('files-selected', {fileHandles})
+      )
+    } else {
+      this.dispatchEvent(Bubble<void>('cancelled', undefined))
+    }
+  }
+
+  /**
+   * Handles the <input type=file> field as a fallback
+   */
+  #processFileInput = async (
+    ev: InputEvent & {currentTarget: HTMLInputElement}
+  ): Promise<void> => {
+    if (ev.currentTarget.files) {
+      await this.#processFileList(ev.currentTarget.files)
+    }
+  }
+
+  /**
+   * Handles the drag'n'drop result
+   */
+  #processDrop = async (ev: InputEvent & {currentTarget: HTMLInputElement}) => {
+    ev.preventDefault()
+    this.#dragEnd()
+
+    if (!ev.dataTransfer || !this.#validateFileList(ev.dataTransfer.files)) {
+      return
+    }
+
+    if (hasFileAccessAPI) {
+      const fileHandles: FileSystemFileHandle[] = []
+      for (let index = 0; index < ev.dataTransfer.items.length; index++) {
+        const item = ev.dataTransfer.items[index]
+        if (item) {
+          const handle = await tryGetAsFilesystemHandle(item)
+          if (handle) {
+            if (handle.kind === 'file') {
+              fileHandles.push(handle)
+            } else {
+              this._errorMessage = 'Directories cannot be dropped here.'
+              return
+            }
+          }
+        }
+      }
+      this.dispatchEvent(
```

Comment:
:no_bell: do we care if fileHandles is empty?

Reply:
It shouldn't be empty by this point.  `#validateFileList()` calls `#validateFileCount()` to ensure there's 1 (if `multiple=false` or 1+ (if `multiple=true`)